### DISCUSSION TITLE
Documentation/resources regarding implementation multi tenancy in KeyCloak

### DISCUSSION THREAD

**Discussion by mcheguini** [keycloak/keycloak#15353] [Q&A]
Recently I was lucky enough to start using KeyCloak,

I am trying to implement multi tenancy using KeyCloak, I want to know which of the following approach works better:

- Each organization (tenant) has its own realm? is there any limitation to creating realms?
- Use a single realm to implement the multi tenancy?

Is there any resource/documentation regarding best practices how to implement multi tenancy in KeyCloak? if so, could you please share with me. I really appreciate it. 
Thank you all  for you supporting KeyCloak (:

**Comment by kkcmadhu** [comment]
i would suggest 1 realm per org, if you are planning to host a SaaS application.
this will keep users of each of your tenant in its own realm, with some tweaking, i am able to serve about 400 500 realm per keycloak safely.
I have tried upto 1000 realm (but keycloak admin console loading etc can be channeling once you have more thn 300 realms), if your typical use case is token issuance to various tenants, this will still scale.

beyond 500 realms,  you may need another keycloak cluster to do the same.. etc

I use a [front proxy pattern (envoy proxy )](https://www.envoyproxy.io/docs/envoy/v1.24.0/start/sandboxes/front_proxy.html?highlight=front%20proxy)  in front of multiple keycloak deployments to scale it beyond 500 realms.

**Comment by xgp** [comment]
Another approach would be to use this extension that adds an Organization model to Keycloak:
https://github.com/p2-inc/keycloak-orgs
https://keycloak.discourse.group/t/keycloak-multi-tenancy-extensions-for-saas-applications/15426
